### ISSUE TITLE
Infix attr accessor

### ISSUE THREAD

**Issue by hadley** [hadley/purrr#69]
Maybe purrr would be a good place for:

```R
`%@%` <- function(x, name) attr(x, name)
```

Need to think about if this should be S3 generic or not, and need to check if this works as is for S4.

**Comment by lionel-** [comment]
I've always been frustrated that `@` does not work as an attribute accessor... Especially since Kevin Ushey mentioned on twitter that S4 slots are actually attributes.

I suppose it would be highly inappropriate to force the matter and export a redefinition?

```r
`@` <- function(x, slot) {
  slot <- as.character(substitute(slot))

  if (isS4(x)) {
    env <- new.env(parent = parent.frame())
    env$`@` <- base::`@`
    eval(call("@", x, slot), envir = env)
  } else {
    attr(x, slot)
  }
}

mtcars@names
```

**Comment by hadley** [comment]
Yeah, we can't override existing functions, so I think `%@%` is the best we can do.  It might be worth trying to push upstream to R core though.

**Comment by nicolasCoutin** [comment]
would this follow then?

```r
x %@% "foo" <- "bar"    # attr(x,"foo") <- "bar"     // x.attr("foo") = "bar";
```
OR use `%<@%` analogous to the `%<>%`  relationship `%>%`

if so then what of
 ```r
attributes(x) <- list(mycomment = "really special", foo = 3:2,bar = paste(1:6))
```

```r
x %<@% list(mycomment = "really special", foo = 3:2,bar = paste(1:6))
```
then you need an `.append = TRUE` or such...

**Comment by lionel-** [comment]
The first one works:

```r
`%@%` <- attr
`%@%<-` <- `attr<-`

mtcars %@% "names"
mtcars %@% "names" <- 1:11
```

The idea of overloading  these operators to work with lists seems interesting.

```r

`%@%` <- function(x, slots) {
  if (length(slots) > 1) {
    attributes(x)[slots]
  } else {
    attr(x, slots)
  }
}

`%@%<-` <- function(x, slots, value) {
  if (length(slots) > 1) {
    attributes(x)[slots] <- value
    x
  } else {
    attr(x, slots) <- value
  }
}

mat <- as.matrix(mtcars)
mat %@% c("dimnames", "dim")
mat %@% c("dimnames", "dim") <- list(NULL, c(2, 16, 11))
```

But it's not good to have the output type change as a function of the number of slots requested. So we'd have to choose one of those two behaviours. The most useful is the list of attributes accessor probably, and we can still hope that R core will enable single attribute lookup with `@` in the future :)

**Comment by lionel-** [comment]
> it's not good to have the output type change as a function of the number of slots requested.

One possibility is to make the behaviour depend on whether the slot is a vector or a list. But the syntax becomes kind of heavy then
```r
mat %@% list("dimnames", "dim") <- list(NULL, c(2, 16, 11))
```

**Comment by hadley** [comment]
I'd rather keep it simple and stick with just getting of single values.

**Comment by hadley** [comment]
Closed in 97ddec7
